Automatic Python functional and linter tests

kamalq97 opened the pull request:
Under the `.github/workflows` directory two `*.yaml` files have been added to carry out functional tests (using `test_*.py` files in the directory `myapp/tests/`) and run linter tests (on *.py files outside `myapp/tests/`)

These tests were made to run automatically each time a new pull request is opened.

kamalq97 committed:
Add unit test Shell scripts and GitHub workflow

kamaql97 commented:
@murradkh the two tests failed (access to files denied), can you maybe edit the file permissions of the two `*.sh` file in Linux and see if it works?

tommerw commented on func_tests.sh:
```diff
@@ -0,0 +1,13 @@
+	python3 -m unittest $testfile
+	exitcode=$(expr  $? \| "$exitcode")
+done
+
```
@kamaql97 you should not have to write a script like this.  I suggest using a tool like [pytest](https://docs.pytest.org/en/stable/) instead.  it easily lets you run all tests in a directory hierarchy with sophisticated filtering, reporting, code coverage etc.

kamaql97 replied:
Thank you for your feedback.  We will look into the use of `pytest` instead. I gravitated towards this solution since this is what was provided to us in [the stage one repository ](https://github.com/kamaql97/python-course-kamal). I just attempted to replicate a tried and tested method that integrates well with GitHub (i.e. it displays the test results within the pull request UI).

kamalq97 replied:
I am now doing it using the GitHub-recommended `.yaml` workflows and I have removed the Linux shell scripts, as per your advice

kamalq97 committed:
Create new GitHub workflow files

These files are based on the Django and pylint workflow templates provided by GitHub (available under the 'Actions' tab on the repository page)

kamalq97 committed:
Delete Linux shell scripts

These scripts are no longer needed since the test workflows have been rewritten

kamalq97 committed:
Update example_test.py

An __init__.py file has also been added to this directory to help Django identify where the tests are located

kamalq97 committed:
Update pylint-workflow.yml

Fatal errors (which will be caught by the Django Test) are excluded from the pylint check, and the tests directory is ignored.

kamalq97 committed:
Add migrations

kamalq97 committed:
Update django-workflow.yml

kamalq97 committed:
Merge branch 'master' into unit_tests

kamalq97 committed:
Update pylint-workflow.yml

tommerw commented:
Thanks @kamaql97 looks good.  However note that the pylint jobs are failing with a bunch of false import errors.  The pylint PYTHONPATH needs to be configured correctly so that it can import all the packages.     since the Django tests are working well I think it would be ok for you to split this ticket up.  You can create one pull request for the functional tests and figure out pylint at your leasure

Also, you definitely need a separate pull request for the "Profile" database migration, that can be reviewed and merged ASAP.  The master branch is broken without it.

murradkhalil committed:
refoctoring unit-testing-github-workflow

murradkh commented on myapp/.pylintrc:
```diff
@@ -0,0 +1,593 @@
+[MASTER]
```
do we need this file kamal or we can add it to ignore file?

kamalq97 replied:
This is the configuration file for the pylint test, we can locate it in a different directory if you prefer it to be more hidden

murradkh commented on .github/workflows/pylint-workflow.yml:
```diff
@@ -0,0 +1,25 @@
+name: Linter Test
```
`pylinting ` i think is not beneficial in our project, due to the fact that's most of the linting issues outputted not related to the code we wrote, which leads to some confusion to recognize which code belongs to us or not in the output; what you think?

kamalq97 replied:
Yeah, we seem to be having a ton of import issues. We need to find a way of importing the Django modules into the linter test.

murradkh commented on .github/workflows/unit-testing-workflow.yml:
```diff
@@ -0,0 +1,31 @@
+name: Unit Testing
```
I added some refactors, like renamings and dependencies-caching

murradkhalil committed:
resolve conflict

murradkhalil committed:
fix

murradkhalil committed:
removing pylinting

murradkhalil committed:
remove pylintrc
